src/passes/validate.rs

```diff
@@ -199,12 +215,42 @@ impl<'a> Visitor<'a> for Validate<'a> {
     }
 
     fn visit_load(&mut self, e: &Load) {
-        self.memarg(&e.arg, e.kind.width());
+        if e.kind.atomic() {
+            self.require_atomic(e.memory, &e.arg, e.kind.width());
+        } else {
+            self.memarg(&e.arg, e.kind.width());
+        }
         e.visit(self);
     }
 
     fn visit_store(&mut self, e: &Store) {
-        self.memarg(&e.arg, e.kind.width());
+        if e.kind.atomic() {
+            self.require_atomic(e.memory, &e.arg, e.kind.width());
+        } else {
+            self.memarg(&e.arg, e.kind.width());
+        }
+        e.visit(self);
+    }
+
+    fn visit_atomic_rmw(&mut self, e: &AtomicRmw) {
+        self.require_atomic(e.memory, &e.arg, e.width.bytes());
+        e.visit(self);
+    }
+
+    fn visit_cmpxchg(&mut self, e: &Cmpxchg) {
+        self.require_atomic(e.memory, &e.arg, e.width.bytes());
+        e.visit(self);
+    }
+
+    fn visit_atomic_notify(&mut self, e: &AtomicNotify) {
+        self.require_shared(e.memory);
+        // TODO: should alignment or things be validated?
```

Comment:
FWIW I think the spec is somewhat unclear on this, so I figured we can just always add points later as necessary